lib/text/ui_text_displayer.js

```diff
@@ -382,15 +382,42 @@ shaka.text.UITextDisplayer = class {
       // TODO: Revisit this as we update the TTML parser.  See special case for
       // backgroundImage + region above.
       const percentageUnit = shaka.text.CueRegion.units.PERCENTAGE;
-      const heightUnit = cue.region.heightUnits == percentageUnit ? '%' : 'px';
-      const widthUnit = cue.region.widthUnits == percentageUnit ? '%' : 'px';
-      const viewportAnchorUnit =
-          cue.region.viewportAnchorUnits == percentageUnit ? '%' : 'px';
-      style.height = cue.region.height + heightUnit;
-      style.width = cue.region.width + widthUnit;
+
+      // Convert everything to percentages of the video container.
+      let width = cue.region.width;
+      let height = cue.region.height;
+      let viewportAnchorX = cue.region.viewportAnchorX;
+      let viewportAnchorY = cue.region.viewportAnchorY;
+      const videoContainerWidth = this.videoContainer_.clientWidth;
```

Comment:
As far as I could tell from another part of this file and from a quick search online, this is the right way to get the width of the video container. I would like to get a set of eyes on this, and want to verify this is the best way to do it.

Reply:
There are many ways to do this, unfortunately.  And to make matters worse, it seems that there are possible performance implications for all of them.

MDN has a neat article on the different ways to get the info.  See if any of these is a better fit than the others: https://developer.mozilla.org/en-US/docs/Web/API/CSS_Object_Model/Determining_the_dimensions_of_elements

One thing I noticed is that clientWidth, offsetWidth, and scrollWidth all round to integers.  getClientBoundingRect() does not round to integer pixels.  (This is what we use in our screenshot tests to get the most accurate and stable screenshots possible.)  How can pixels not be integers, you might ask?  I had no idea until a few months back working on screenshot tests, but browsers often do "sub-pixel rendering", where the graphics buffer they write to internally is actually a different size than what's ultimately shown to the user.

This quora article mentions "layout thrashing" caused by reading these values.  Essentially, just by asking the width, you can force the browser to re-calculate the layout of the whole page, which is expensive: https://www.quora.com/How-expensive-is-getBoundingClientRect-in-a-JavaScript-request-animation-frame

This Google article discusses what makes this happen, and ways to avoid it: https://developers.google.com/web/fundamentals/performance/rendering/avoid-large-complex-layouts-and-layout-thrashing

In short, if you do all your reads and then all your writes, you won't trigger expensive layout checks in the browser multiple times.

This gist details what things can trigger a layout of the page if there have been any writes since your last read: https://gist.github.com/paulirish/5d52fb081b3570c81e3a

Reply:
Thanks so much for the very detailed answer! These are interesting reads - I was not aware of sub-pixels either. Keeping all of this in mind, I think in this case clientWidth is suitable for our needs, since precision to the nearest pixel is good enough and the very slight loss of precision is probably better than the expensive bounding rect operation. I'll keep it for now, unless you think getting the bounded rect is a better solution.

Reply:
getClientBoundingRect() and clientWidth can both be expensive from what I've read today.  If you've changing anything about the page between reads to clientWidth, you can trigger a layout.  So it would be best to read that once, outside the loop that manipulates the DOM for each cue.